### PR TITLE
Ncnews 14

### PR DESCRIPTION
**Pull request by abdullah7x**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -228,3 +228,90 @@ describe('POST /api/articles/:article_id/comments', () => {
...
+    });
+    const res2 = await request(app)
+      .get('/api/articles?sort_by=comment_count&order=asc')
+      .expect(200);
+    expect(res2.body).toBeSortedBy('comment_count', {
+      coerce: true,
+      ascending: true,
+    });
+  });
+  test('200: returns in order of choice sorted by created_at when given no column', async () => {
```
Comment (Rosem10) [inline __tests__/app.test.js:269]: Could you please rephrase this to reference "sorted_by" instead of the column specifically? We shouldn't start giving away clues about our database in our tests.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -228,3 +228,90 @@ describe('POST /api/articles/:article_id/comments', () => {
...
+      coerce: true,
+      ascending: true,
+    });
+    const res2 = await request(app).get('/api/articles?order=desc').expect(200);
+    expect(res.body).toBeSortedBy('created_at', {
+      coerce: true,
+      descending: true,
+    });
+  });
+  test('200: returns articles on a given topic', async () => {
```
Comment (Rosem10) [inline __tests__/app.test.js:281]: Please test for if passed a topic that exists but has no associated articles.

**Review by Rosem10 — CHANGES_REQUESTED** [review]
Just a couple of little changes to be made please!

File: models/articles.models.js [hunk 3]
```diff
@@ -40,14 +41,50 @@ exports.checkArticleExists = async (article_id) => {
...
+    'created_at',
+    'votes',
+    'comment_count',
+  ];
+  if (!validColumns.includes(sort_by)) {
+    return Promise.reject({ status: 400, message: 'bad request' });
+  }
+  const queryValues = [];
+  let queryStr = `SELECT articles.*, COUNT(comment_id) AS comment_count
+  FROM articles
```
Comment (xmenbteam) [inline models/articles.models.js:72]: How can we ensure that the comment_count will always be an INT? Look up CASTing (`::`) in SQL.

File: models/articles.models.js [hunk 4]
```diff
@@ -40,14 +41,50 @@ exports.checkArticleExists = async (article_id) => {
...
+
+  if (topic) {
+    queryStr += ` WHERE topic = $1`;
+    queryValues.push(topic);
+  }
+
+  queryStr += ` GROUP BY articles.article_id 
+  ORDER BY ${sort_by}`;
+
+  if (/^desc$/i.test(order) || /^asc$/i.test(order)) {
```
Comment (xmenbteam) [inline models/articles.models.js:84]: This is nice! But you should be rejecting invalid order_by values BEFORE you build your query string, probably on line 69.

**Review by xmenbteam — CHANGES_REQUESTED** [review]
Just a couple of things to change here. Your code looks really good though! Well written :)

File: models/articles.models.js (hunk 3 above)
Comment (abdullah7x) [reply models/articles.models.js:72]: I can't figure out how to use CAST alongside the aggregate function and alias

File: __tests__/app.test.js [hunk 5]
```diff
@@ -228,3 +228,94 @@ describe('POST /api/articles/:article_id/comments', () => {
...
+    const res = await request(app).get('/api/articles?topic=mitch').expect(200);
+    res.body.forEach((article) => {
+      expect(article.topic).toBe('mitch');
+    });
+  });
+  test('200: returns an empty array when given a topic which exists but has no articles', async () => {
```
Comment (dentednerd) [inline __tests__/app.test.js:287]: Are you sure that this should return a 200 status?

File: models/articles.models.js [hunk 6]
```diff
@@ -40,14 +41,49 @@ exports.checkArticleExists = async (article_id) => {
...
+exports.selectAllArticles = async (
+  sort_by = 'created_at',
+  order = 'desc',
+  topic
+) => {
+  const topics = await selectAllTopics();
+  topicsArray = [];
+  topics.forEach((object) => {
+    topicsArray.push(object.slug);
+  });
```
Comment (dentednerd) [inline models/articles.models.js:53]: This is a longwinded way of getting an array of topic slugs. Think of a better array method you could use here. I bet you can get this down to one line!

File: models/articles.models.js (hunk 3 above)
Comment (dentednerd) [reply models/articles.models.js:75]: Use nchelp!

**Review by dentednerd — CHANGES_REQUESTED** [review]
A few things to think about here.

File: __tests__/app.test.js (hunk 5 above)
Comment (abdullah7x) [reply __tests__/app.test.js:287]: My thinking was it isn't a bad request and it isn't 'not found' either technically because the topic exists, so the client can see that the topic exists but there aren't any existing articles atm. Correct me if I'm wrong

Comment (dentednerd) [reply __tests__/app.test.js:287]: Remember how we were saying that a 200 status is like finding a box with contents, but that a 204 status is for finding a box with nothing inside of it?
